components/shared/compositing/viewport_description.rs

```diff
@@ -0,0 +1,171 @@
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum UserScalable {
+    /// Zoom is not allowed
+    No = 0,
+    /// Zoom is allowed
+    Yes = 1,
+}
+
+/// Parses a viewport user scalable value.
+impl TryFrom<&str> for UserScalable {
+    type Error = &'static str;
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        match value.to_lowercase().as_str() {
+            "yes" => Ok(UserScalable::Yes),
+            "no" => Ok(UserScalable::No),
+            _ => match value.parse::<f32>() {
+                Ok(1.0) => Ok(UserScalable::Yes),
+                Ok(0.0) => Ok(UserScalable::No),
+                _ => Err("can't convert character to UserScalable"),
+            },
+        }
+    }
+}
+
+impl Default for ViewportDescription {
+    fn default() -> Self {
+        ViewportDescription {
+            initial_scale: Scale::new(DEFAULT_ZOOM),
+            minimum_scale: Scale::new(MIN_ZOOM),
+            maximum_scale: Scale::new(MAX_ZOOM),
+            user_scalable: UserScalable::Yes,
+        }
+    }
+}
+
+impl ViewportDescription {
+    /// Iterates over the key-value pairs generated from meta tag and returns a ViewportDescription
+    fn process_viewport_key_value_pairs(pairs: HashMap<String, String>) -> ViewportDescription {
+        let mut description = ViewportDescription::default();
+        for (key, value) in &pairs {
+            match key.as_str() {
+                "initial-scale" => {
+                    if let Some(zoom) = Self::parse_viewport_value_as_zoom(value) {
+                        description.initial_scale = zoom;
+                    }
+                },
+                "minimum-scale" => {
+                    if let Some(zoom) = Self::parse_viewport_value_as_zoom(value) {
+                        description.minimum_scale = zoom;
+                    }
+                },
+                "maximum-scale" => {
+                    if let Some(zoom) = Self::parse_viewport_value_as_zoom(value) {
+                        description.maximum_scale = zoom;
+                    }
+                },
+                "user-scalable" => {
+                    if let Ok(user_zoom_allowed) = value.as_str().try_into() {
+                        description.user_scalable = user_zoom_allowed;
+                    }
+                },
+                _ => (),
+            }
+        }
+        description
+    }
+
+    /// Parses a viewport zoom value.
+    fn parse_viewport_value_as_zoom(value: &str) -> Option<Scale<f32>> {
+        value
+            .to_lowercase()
+            .as_str()
+            .parse::<f32>()
+            .ok()
+            .filter(|&n| (0.0..=10.0).contains(&n))
+            .map(Scale::new)
+    }
+
+    /// Constrains a zoom value within the allowed scale range
+    pub fn clamp_zoom(&self, zoom: f32) -> f32 {
+        zoom.clamp(self.minimum_scale.get(), self.maximum_scale.get())
+    }
+}
+
+/// <https://drafts.csswg.org/css-viewport/#parsing-algorithm>
```

Comment:
I'm just thinking something along the lines of this:

```suggestion
/// <https://drafts.csswg.org/css-viewport/#parsing-algorithm>
///
/// This implementation differs from the specified algorithm, but is equivalent because
/// ...explanation of how operations below are equivalent to spec algorithm...
```

Reply:
I added some comments, Please take a look.

Reply:
Thanks, this is a lot better!